driver-core/src/test/unit/com/mongodb/ConnectionStringSpecification.groovy

```diff
@@ -131,6 +124,35 @@ class ConnectionStringSpecification extends Specification {
         type << ['amp', 'semi', 'mixed']
     }
 
+    @Unroll
```

Comment:
The majority of the change is to ensure we fail when its expected and that we provide helpful error messages.  

Other behavioural changes are warning on deprecated values, and explicit failing for:  invalid credentials and unsupported or invalid hosts.